### PR TITLE
custom autograd func memory

### PR DESCRIPTION
**Pull request by pengwa**
**Description**: Refine memory usage of custom autograd function fallback.

This PR mainly targets to 
1. remove PythonOpGrad's control dependency upon its corresponding PythonOp's outputs. 
2. Fix the segement fault once doing 1.

With this change, just a few MB is saved. So we still have a gap of 12512.0015 MB V.S. 10016.001 MB.

    ORT Before:
    Step 19 MA 9312.001 MB         Max_MA 12544.0015 MB         CPU Virtual Memory:  used = 54.43 GB, percent = 12.3%
    
    ORT with the change:
    Step 19 MA 9312.001 MB         Max_MA 12512.0015 MB         CPU Virtual Memory:  used = 57.39 GB, percent = 13.0%
    
    
    PT: Step 19 MA 9312.0005 MB         Max_MA 10016.001 MB         CPU Virtual Memory:  used = 53.86 GB, percent = 12.2%

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 1]
```diff
@@ -1785,10 +1785,9 @@ IMPLEMENT_GRADIENT_BUILDER(GetPythonOpGradient) {
   }
 
   // Also connect forward outputs to PythonOpGrad for random segement fault issues.
-  // Todo (pengwa): we should investigate whether we could avoid those outputs that are not used
-  // in backward computation.
+  // Todo (pengwa): remove the control dependency from PythonOpGrad schema.
   for (int i = 1; i < GetSrcNodeOutputSize(); ++i) {
-    input_args.push_back(O(i));
+    input_args.push_back(ArgDef());
```
Comment (wschin) [inline orttraining/orttraining/core/graph/gradient_builder.cc:1790]: Is this loop still needed?

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/torch_interop_utils/torch_interop_utils.cc [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    static PyNodeSharedPointerPool pool;
+    return pool;
+  };
+
+  void RegisterGradFunc(const size_t& ctx_address, torch::autograd::AutogradMeta* autograd_meta){
+    auto it = grad_fns_.find(ctx_address);
+    TORCH_CHECK(it == grad_fns_.end(), "should not register grad_fn twice for ctx ", ctx_address);
+
+    // Add new entry if key hasn't been registered.
+    grad_fns_.emplace(ctx_address, std::move(autograd_meta->grad_fn_));
```
Comment (wschin) [inline orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/torch_interop_utils/torch_interop_utils.cc:35]: Does it mean we steal the ownership of `grad_fn_` from Pytorch? I thought we just need to add one reference count to an `PyObject`.

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 1 above)
Comment (pengwa) [reply orttraining/orttraining/core/graph/gradient_builder.cc:1790]: Yes, We can remove them. and those control inputs in PythonOpGrad Op schema. We can split the changes for easier review I think.

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/torch_interop_utils/torch_interop_utils.cc (hunk 2 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/torch_interop_utils/torch_interop_utils.cc:35]: grad_fn_  we got from Python side is a THPFunction*, which is owned by PyNode. 

The real gradient function torch runs is the PyNode. So if we want the ctx live, we need make sure the PyNode live. The way we make the PyNode live is, increate a reference count here to the std::shared_ptr<PyNode>.

**Review by wschin — DISMISSED** [review]
LGTM!

**Review by SherlockNoMad — APPROVED** [review]

**Comment by pengwa** [comment]
Thank you @wschin and @SherlockNoMad !!!
